keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,585 @@
+cycle.
+
+1. The Job controller calculates the number of succeeded Pods as the sum of:
+   - `.status.succeeded`,
+   - the size of `job.status.uncountedTerminatedPods.succeeded` and
+   - the number of finished Pods that are not in `job.status.uncountedTerminatedPods.succeeded`.
+   This number informs the creation of missing Pods to reach `.spec.completions`.
+   The controller creates Pods for a Job with the finalizer
+   `batch.kubernetes.io/job-completion`.
+2. The Job controller adds Pod UIDs to the `.status.uncountedTerminatedPods.succeeded`
+   and `.status.uncountedTerminatedPods.failed` lists if the Pod:
+    - has the `batch.kubernetes.io/job-completion` finalizer, and
+    - the Pod is on Succeeded or Failed phase, respectively.
+    The controller sends a status update.
+3. The Job controller removes the `batch.kubernetes.io/job-completion` finalizer
+   from all Pods on Succeeded or Failed phase that were added to the lists in
+   `.status.uncountedTerminatedPods` in the previous step.
+4. The Job controller counts the Pods in the `.status.uncountedTerminatedPods` lists
+   that:
+   - have no finalizer, or
+   - were removed from the system.
+   The counts increment the `.status.failed` and `.status.succeeded` and clears
+   counted Pods from `.status.uncountedPodsUIDs` lists. The controller sends a
+   status update.
+
+Steps 2 to 4 might deal with a potentially big number of Pods. Thus, status
+updates can potentially stress the kube-apiserver. For this reason, the Job
+controller repeats steps 2 to 4, capping the number of Pods each time, until
+the controller processes all the Pods. The number of Pods is caped by:
+- time: in each iteration, the job controller removes all the Pods' finalizers
+  it can in a unit of time in the order of tens of seconds. This allows to
+  throttle the number of Job status updates.
+- count: Preventing big writes to etcd. We limit the number of UIDs to the order
+  of hundreds, keeping the size of the slice under 20kb.
+
+If any Pod finalizer removal fails in step 3, the controller manager still
+executes step 4 with the Pods that succeeded.
+
+Steps 2 to 4 might be skipped in the scenario where a status update happened
+too recently and the number of uncounted Pods is a small percentage of
+parallelism.
+
+### Deleted Pods
+   
+In the case where a user or another controller removes a Pod, which sets a
+deletion timestamp, the Job controller treats it the same as any other Pod.
+That is, once it reaches Failed status, the controller accounts for the Pod and
+then removes the finalizer.
+
+This is different from the legacy tracking, where the Job controller does not
+account for deleted Pods. This is a limitation that this KEP also wants to
+solve.
+
+However, if the Job controller deletes the Pod (when parallelism is decreased,
+for example), the controller removes the finalizer before deleting it. Thus,
+these deletions don't count towards the failures.
```

Comment:
From the description here, it'd apply to suspended Jobs ([KEP 2232](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/2232-suspend-jobs/)) as well, I assume?

Reply:
yes, but applies to any other future feature that might delete pods.